File: apk/app/src/main/java/com/example/momentra/ui/auth/AuthViewModel.kt
package com.example.momentra.ui.auth

import android.app.Activity
import android.app.Application
import android.content.Context
import androidx.lifecycle.AndroidViewModel
import androidx.lifecycle.viewModelScope
import com.example.momentra.analytics.MomentraAnalytics
import com.example.momentra.data.api.ApiClient
import com.example.momentra.data.api.ApiResultException
import com.example.momentra.data.auth.AuthErrorMapper
import com.example.momentra.data.auth.FirebaseAuthRepository
import com.example.momentra.data.auth.PhoneSendResult
import com.example.momentra.data.device.DeviceRegistrar
import com.example.momentra.data.local.AppPreferences
import com.example.momentra.data.repository.MeRepository
import com.example.momentra.data.security.SecurityPreferences
import com.example.momentra.domain.AuthPhase
import com.example.momentra.domain.ShellIdentity
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.flow.update
import kotlinx.coroutines.launch

data class AuthUiState(
    val phase: AuthPhase = AuthPhase.Launching,
    val identity: ShellIdentity? = null,
    val phoneCodeSent: Boolean = false,
    val error: String? = null,
) {
    val isLoggedIn: Boolean get() = phase == AuthPhase.Authenticated && identity != null
    val isRestoring: Boolean get() = phase == AuthPhase.RestoringSession || phase == AuthPhase.AuthenticatedBootstrapping
    val isLoading: Boolean get() = phase == AuthPhase.Authenticating || phase == AuthPhase.AuthenticatedBootstrapping
}

class AuthViewModel @JvmOverloads constructor(
    application: Application,
    private val authRepository: FirebaseAuthRepository = FirebaseAuthRepository(),
    private val meRepository: MeRepository = MeRepository(application),
) : AndroidViewModel(application) {

    private val prefs = AppPreferences(application)
    private val _state = MutableStateFlow(AuthUiState())
    val state: StateFlow<AuthUiState> = _state.asStateFlow()

    init {
        try {
            restoreSessionIfPossible()
        } catch (e: Exception) {
            _state.update {
                it.copy(phase = AuthPhase.SignedOut, error = e.message)
            }
        }
    }

    private fun restoreSessionIfPossible() {
        if (authRepository.currentUser == null) {
            _state.update { it.copy(phase = AuthPhase.SignedOut) }
            return
        }
        _state.update { it.copy(phase = AuthPhase.RestoringSession) }
        viewModelScope.launch { bootstrapAfterAuth(restore = true) }
    }

    private suspend fun bootstrapAfterAuth(restore: Boolean = false) {
        _state.update {
            it.copy(
                phase = if (restore) AuthPhase.RestoringSession else AuthPhase.AuthenticatedBootstrapping,
                error = null,
            )
        }
        meRepository.getMe().fold(
            onSuccess = { identity ->
                authRepository.currentUser?.uid?.let { firebaseUid ->
                    prefs.saveCachedIdentity(
                        firebaseUid = firebaseUid,
                        userId = identity.userId,
                        displayName = identity.displayName,
                        email = identity.email,
                    )
                }
                MomentraAnalytics.get().syncUserDemographics(
                    user = authRepository.currentUser,
                    profileDisplayName = identity.displayName,
                    profileEmail = identity.email,
                )
                if (!restore) {
                    MomentraAnalytics.get().trackAuthResult("bootstrap", success = true)
                }
                ApiClient.warmAuthToken()
                viewModelScope.launch { DeviceRegistrar.register(getApplication()) }
                _state.update {
                    AuthUiState(
                        phase = AuthPhase.Authenticated,
                        identity = identity,
                        phoneCodeSent = false,
                        error = null,
                    )
                }
            },
            onFailure = { e ->
                when (e) {
                    is ApiResultException.Network -> {
                        // Offline: keep Firebase session; never substitute Firebase UID for Momentra userId.
                        if (restore && authRepository.currentUser != null) {
                            val firebaseUid = authRepository.currentUser?.uid
                            val cached = firebaseUid?.let { prefs.getCachedIdentity(it) }
                            if (cached != null) {
                                _state.update {
                                    AuthUiState(
                                        phase = AuthPhase.Authenticated,
                                        identity = ShellIdentity(
                                            userId = cached.first,
                                            displayName = cached.second,
                                            email = cached.third,
                                            firebaseUid = firebaseUid,
                                        ),
                                        error = "NETWORK_UNAVAILABLE",
                                    )
                                }
                            } else {
                                _state.update {
                                    AuthUiState(
                                        phase = AuthPhase.RestoringSession,
                                        identity = null,
                                        error = "NETWORK_UNAVAILABLE",
                                    )
                                }
                            }
                        } else {
                            failBootstrap(e, restore)
                        }
                    }
                    is ApiResultException.Unauthenticated -> {
                        authRepository.signOut()
                        _state.update {
                            AuthUiState(phase = AuthPhase.SessionExpired, error = e.message)
                        }
                    }
                    else -> failBootstrap(e, restore)
                }
            },
        )
    }

    private fun failBootstrap(e: Throwable, restore: Boolean) {
        authRepository.signOut()
        if (!restore) {
            MomentraAnalytics.get().trackAuthResult("bootstrap", success = false, errorCode = e.message)
        }
        _state.update {
            AuthUiState(
                phase = if (restore) AuthPhase.SignedOut else AuthPhase.AuthError,
                error = if (restore) null else (e.message ?: "Bootstrap failed"),
            )
        }
    }

    fun onFirebaseLoggedIn() {
        _state.update { it.copy(phase = AuthPhase.AuthenticatedBootstrapping, error = null) }
        viewModelScope.launch { bootstrapAfterAuth() }
    }

    fun onAuthStarted() {
        _state.update { it.copy(phase = AuthPhase.Authenticating, error = null) }
    }

    fun onAuthError(message: String) {
        _state.update {
            it.copy(phase = AuthPhase.AuthError, error = AuthErrorMapper.userMessage(message))
        }
    }

    fun clearError() {
        _state.update { it.copy(error = null) }
    }

    fun signInWithEmail(email: String, password: String) {
        _state.update { it.copy(phase = AuthPhase.Authenticating, error = null) }
        viewModelScope.launch {
            authRepository.signIn(email, password).fold(
                onSuccess = {
                    MomentraAnalytics.get().trackAuthResult("email_sign_in", success = true)
                    onFirebaseLoggedIn()
                },
                onFailure = { e ->
                    MomentraAnalytics.get().trackAuthResult("email_sign_in", success = false, errorCode = e.message)
                    _state.update {
                        it.copy(
                            phase = AuthPhase.AuthError,
                            error = AuthErrorMapper.userMessage(e.message ?: "Sign-in failed"),
                        )
                    }
                },
            )
        }
    }

    fun registerWithEmail(email: String, password: String, confirmPassword: String) {
        if (password != confirmPassword) {
            _state.update { it.copy(error = "Passwords do not match.") }
            return
        }
        _state.update { it.copy(phase = AuthPhase.Authenticating, error = null) }
        viewModelScope.launch {
            authRepository.register(email, password).fold(
                onSuccess = {
                    MomentraAnalytics.get().trackAuthResult("email_register", success = true)
                    onFirebaseLoggedIn()
                },
                onFailure = { e ->
                    MomentraAnalytics.get().trackAuthResult("email_register", success = false, errorCode = e.message)
                    _state.update {
                        it.copy(
                            phase = AuthPhase.AuthError,
                            error = AuthErrorMapper.userMessage(e.message ?: "Registration failed"),
                        )
                    }
                },
            )
        }
    }

    fun sendPhoneCode(activity: Activity, phone: String) {
        _state.update { it.copy(phase = AuthPhase.Authenticating, error = null) }
        viewModelScope.launch {
            authRepository.sendPhoneCode(activity, phone).fold(
                onSuccess = { result ->
                    when (result) {
                        is PhoneSendResult.AutoVerified -> {
                            MomentraAnalytics.get().trackAuthResult("phone_auto", success = true)
                            onFirebaseLoggedIn()
                        }
                        PhoneSendResult.CodeSent -> {
                            MomentraAnalytics.get().trackAuthResult("phone_sms_sent", success = true)
                            _state.update {
                                it.copy(phase = AuthPhase.SignedOut, phoneCodeSent = true, error = null)
                            }
                        }
                    }
                },
                onFailure = { e ->
                    MomentraAnalytics.get().trackAuthResult("phone_sms_sent", success = false, errorCode = e.message)
                    _state.update {
                        it.copy(
                            phase = AuthPhase.AuthError,
                            error = AuthErrorMapper.userMessage(e.message ?: "Could not send SMS code"),
                        )
                    }
                },
            )
        }
    }

    fun confirmPhoneCode(code: String) {
        if (code.trim().length < 6) {
            _state.update { it.copy(error = "Enter the 6-digit code from SMS.") }
            return
        }
        _state.update { it.copy(phase = AuthPhase.Authenticating, error = null) }
        viewModelScope.launch {
            authRepository.confirmPhoneCode(code).fold(
                onSuccess = {
                    MomentraAnalytics.get().trackAuthResult("phone_verify", success = true)
                    onFirebaseLoggedIn()
                },
                onFailure = { e ->
                    MomentraAnalytics.get().trackAuthResult("phone_verify", success = false, errorCode = e.message)
                    _state.update {
                        it.copy(
                            phase = AuthPhase.AuthError,
                            error = AuthErrorMapper.userMessage(e.message ?: "Phone verification failed"),
                        )
                    }
                },
            )
        }
    }

    fun resetPhoneFlow() {
        authRepository.resetPhoneFlow()
        _state.update { it.copy(phoneCodeSent = false, error = null, phase = AuthPhase.SignedOut) }
    }

    fun sendPasswordReset(email: String) {
        if (email.isBlank()) {
            _state.update { it.copy(error = "Enter your email to reset password.") }
            return
        }
        _state.update { it.copy(phase = AuthPhase.Authenticating, error = null) }
        viewModelScope.launch {
            authRepository.sendPasswordReset(email).fold(
                onSuccess = {
                    _state.update {
                        it.copy(
                            phase = AuthPhase.SignedOut,
                            error = "Password reset email sent. Check your inbox.",
                        )
                    }
                },
                onFailure = { e ->
                    _state.update {
                        it.copy(
                            phase = AuthPhase.AuthError,
                            error = AuthErrorMapper.userMessage(e.message ?: "Could not send reset email"),
                        )
                    }
                },
            )
        }
    }

    fun signInWithGoogle(context: Context) {
        _state.update { it.copy(phase = AuthPhase.Authenticating, error = null) }
        viewModelScope.launch {
            authRepository.signInWithGoogle(context).fold(
                onSuccess = {
                    MomentraAnalytics.get().trackAuthResult("google", success = true)
                    onFirebaseLoggedIn()
                },
                onFailure = { e ->
                    MomentraAnalytics.get().trackAuthResult("google", success = false, errorCode = e.message)
                    _state.update {
                        it.copy(
                            phase = AuthPhase.AuthError,
                            error = AuthErrorMapper.userMessage(e.message ?: "Google sign-in failed"),
                        )
                    }
                },
            )
        }
    }

    fun signOut() {
        val firebaseUid = authRepository.currentUser?.uid
        val momentraUserId = _state.value.identity?.userId
        prefs.clearCachedIdentity(firebaseUid)
        prefs.clearUserScopedShell(momentraUserId)
        meRepository.clearBootstrapCache(momentraUserId)
        SecurityPreferences(getApplication()).clearUserScoped(momentraUserId)
        authRepository.signOut()
        _state.value = AuthUiState(phase = AuthPhase.SignedOut)
    }

    fun onSessionExpired() {
        val firebaseUid = authRepository.currentUser?.uid
        val momentraUserId = _state.value.identity?.userId
        prefs.clearCachedIdentity(firebaseUid)
        meRepository.clearBootstrapCache(momentraUserId)
        authRepository.signOut()
        _state.value = AuthUiState(phase = AuthPhase.SessionExpired, error = "Session expired")
    }
}

File: momentra/momentra/AuthViewModel.swift
import Combine
import FirebaseAuth
import FirebaseCore
import Foundation
import UIKit
#if os(iOS)
import AuthenticationServices
import CryptoKit
import GoogleSignIn
import Security
#endif

@MainActor
final class AuthViewModel: ObservableObject {
    @Published var phase: AuthPhase = .launching
    @Published var identity: ShellIdentity?
    @Published var error: String?
    @Published var phoneCodeSent = false

    var isLoggedIn: Bool { phase == .authenticated && identity != nil }
    var isLoading: Bool { phase == .authenticating || phase == .authenticatedBootstrapping }
    var isRestoringSession: Bool { phase == .restoringSession || phase == .authenticatedBootstrapping }

    private let api = APIClient.shared
    private var phoneVerificationID: String?
#if os(iOS)
    private var appleSignInNonce: String?
#endif

    init() {
        Task { await restoreSessionIfPossible() }
    }

    func restoreSessionIfPossible() async {
        guard Auth.auth().currentUser != nil else {
            phase = .signedOut
            identity = nil
            return
        }
        phase = .restoringSession
        do {
            let me = try await api.bootstrapMe()
            let firebaseUid = Auth.auth().currentUser?.uid
            identity = ShellIdentity(
                userId: me.userId,
                displayName: me.displayName,
                email: me.email,
                firebaseUid: firebaseUid
            )
            if let firebaseUid {
                MomentraIdentityCache.save(
                    firebaseUid: firebaseUid,
                    userId: me.userId,
                    displayName: me.displayName,
                    email: me.email
                )
            }
            MomentraAnalytics.shared.syncUserDemographics(
                user: Auth.auth().currentUser,
                profileDisplayName: me.displayName,
                profileEmail: me.email
            )
            phase = .authenticated
        } catch let err as APIErrorKind {
            switch err {
            case .network:
                // Never substitute Firebase UID for Momentra userId.
                if let firebaseUid = Auth.auth().currentUser?.uid,
                   let cached = MomentraIdentityCache.load(firebaseUid: firebaseUid) {
                    identity = cached
                    phase = .authenticated
                    error = "NETWORK_UNAVAILABLE"
                } else if Auth.auth().currentUser != nil {
                    identity = nil
                    phase = .restoringSession
                    error = "NETWORK_UNAVAILABLE"
                } else {
                    failRestore()
                }
            case .unauthenticated:
                try? Auth.auth().signOut()
                #if os(iOS)
                GIDSignIn.sharedInstance.signOut()
                #endif
                phase = .sessionExpired
                identity = nil
            default:
                failRestore()
            }
        } catch {
            failRestore()
        }
    }

    private func failRestore() {
        try? Auth.auth().signOut()
        #if os(iOS)
        GIDSignIn.sharedInstance.signOut()
        #endif
        phase = .signedOut
        identity = nil
    }

    func signInWithEmailPassword(email: String, password: String) {
        phase = .authenticating
        error = nil
        Task {
            do {
                let result = try await Auth.auth().signIn(
                    withEmail: email.lowercased(),
                    password: password
                )
                try await finishBootstrap(user: result.user, method: "email_sign_in")
            } catch {
                recoverFromFailure(error, method: "email_sign_in")
            }
        }
    }

    func registerWithEmailPassword(email: String, password: String) {
        phase = .authenticating
        error = nil
        Task {
            do {
                let result = try await Auth.auth().createUser(
                    withEmail: email.lowercased(),
                    password: password
                )
                try await finishBootstrap(user: result.user, method: "email_register")
            } catch {
                recoverFromFailure(error, method: "email_register")
            }
        }
    }

    func sendPhoneCode(phone: String) {
        phase = .authenticating
        error = nil
        guard let e164 = Self.normalizedE164(phone) else {
            phase = .authError
            error = "Enter a phone number in E.164 format, e.g. +919876543210."
            return
        }
        Task {
            do {
                let verificationID = try await PhoneAuthProvider.provider().verifyPhoneNumber(e164, uiDelegate: nil)
                phoneVerificationID = verificationID
                phoneCodeSent = true
                phase = .signedOut
                MomentraAnalytics.shared.trackAuthResult(method: "phone_sms_sent", success: true)
            } catch {
                recoverFromFailure(error, method: "phone_sms_sent")
            }
        }
    }

    func confirmPhoneCode(code: String) {
        let trimmed = code.trimmingCharacters(in: .whitespacesAndNewlines)
        guard let verificationID = phoneVerificationID, trimmed.count >= 6 else {
            error = "Enter the 6-digit code from SMS."
            return
        }
        phase = .authenticating
        error = nil
        Task {
            do {
                let credential = PhoneAuthProvider.provider().credential(
                    withVerificationID: verificationID,
                    verificationCode: trimmed
                )
                let result = try await Auth.auth().signIn(with: credential)
                phoneCodeSent = false
                phoneVerificationID = nil
                try await finishBootstrap(user: result.user, method: "phone_verify")
            } catch {
                recoverFromFailure(error, method: "phone_verify")
            }
        }
    }

    func resetPhoneFlow() {
        phoneCodeSent = false
        phoneVerificationID = nil
        error = nil
        phase = .signedOut
    }

    private static func normalizedE164(_ raw: String) -> String? {
        let trimmed = raw.trimmingCharacters(in: .whitespacesAndNewlines)
        if trimmed.hasPrefix("+") {
            let digits = trimmed.dropFirst().filter(\.isNumber)
            return digits.count >= 8 ? "+\(digits)" : nil
        }
        let digits = trimmed.filter(\.isNumber)
        guard digits.count >= 10 else { return nil }
        if digits.count == 10 { return "+91\(digits)" }
        return "+\(digits)"
    }

    #if os(iOS)
    func prepareAppleSignInRequest(_ request: ASAuthorizationAppleIDRequest) {
        let nonce = Self.randomNonceString()
        appleSignInNonce = nonce
        request.requestedScopes = [.fullName, .email]
        request.nonce = Self.sha256(nonce)
        phase = .authenticating
        error = nil
    }

    func handleAppleSignInCompletion(_ result: Result<ASAuthorization, Error>) {
        switch result {
        case .failure(let error):
            if let authError = error as? ASAuthorizationError, authError.code == .canceled {
                phase = .signedOut
                self.error = nil
                appleSignInNonce = nil
                return
            }
            appleSignInNonce = nil
            recoverFromFailure(error, method: "apple")
        case .success(let authorization):
            guard
                let appleIDCredential = authorization.credential as? ASAuthorizationAppleIDCredential,
                let nonce = appleSignInNonce,
                let tokenData = appleIDCredential.identityToken,
                let idToken = String(data: tokenData, encoding: .utf8)
            else {
                appleSignInNonce = nil
                recoverFromFailure(
                    NSError(
                        domain: "MomentraAuth",
                        code: -4,
                        userInfo: [NSLocalizedDescriptionKey: "Apple Sign In returned an incomplete credential."]
                    ),
                    method: "apple"
                )
                return
            }
            appleSignInNonce = nil
            Task {
                do {
                    let credential = OAuthProvider.appleCredential(
                        withIDToken: idToken,
                        rawNonce: nonce,
                        fullName: appleIDCredential.fullName
                    )
                    let authResult = try await Auth.auth().signIn(with: credential)
                    if let fullName = appleIDCredential.fullName {
                        let display = PersonNameComponentsFormatter().string(from: fullName)
                            .trimmingCharacters(in: .whitespacesAndNewlines)
                        if !display.isEmpty, authResult.user.displayName == nil {
                            let change = authResult.user.createProfileChangeRequest()
                            change.displayName = display
                            try? await change.commitChanges()
                        }
                    }
                    try await finishBootstrap(user: authResult.user, method: "apple")
                } catch {
                    recoverFromFailure(error, method: "apple")
                }
            }
        }
    }

    func signInWithGoogle() {
        phase = .authenticating
        error = nil
        Task {
            do {
                guard let clientID = FirebaseApp.app()?.options.clientID else {
                    throw NSError(
                        domain: "MomentraAuth",
                        code: -1,
                        userInfo: [NSLocalizedDescriptionKey: "Missing Firebase client ID"]
                    )
                }
                GIDSignIn.sharedInstance.configuration = GIDConfiguration(clientID: clientID)
                guard let presenter = UIApplication.topViewController() else {
                    throw NSError(
                        domain: "MomentraAuth",
                        code: -2,
                        userInfo: [NSLocalizedDescriptionKey: "Could not present Google sign-in"]
                    )
                }
                let signInResult = try await GIDSignIn.sharedInstance.signIn(withPresenting: presenter)
                guard let idToken = signInResult.user.idToken?.tokenString else {
                    throw NSError(
                        domain: "MomentraAuth",
                        code: -3,
                        userInfo: [NSLocalizedDescriptionKey: "Missing Google ID token"]
                    )
                }
                let credential = GoogleAuthProvider.credential(
                    withIDToken: idToken,
                    accessToken: signInResult.user.accessToken.tokenString
                )
                let authResult = try await Auth.auth().signIn(with: credential)
                try await finishBootstrap(user: authResult.user, method: "google")
            } catch {
                recoverFromFailure(error, method: "google")
            }
        }
    }

    private static func randomNonceString(length: Int = 32) -> String {
        precondition(length > 0)
        let charset = Array("0123456789ABCDEFGHIJKLMNOPQRSTUVXYZabcdefghijklmnopqrstuvwxyz-._")
        var result = ""
        var remaining = length
        while remaining > 0 {
            let randoms: [UInt8] = (0 ..< 16).map { _ in
                var random: UInt8 = 0
                let status = SecRandomCopyBytes(kSecRandomDefault, 1, &random)
                if status != errSecSuccess {
                    fatalError("Unable to generate nonce. SecRandomCopyBytes failed with OSStatus \(status)")
                }
                return random
            }
            for random in randoms where remaining > 0 {
                if random < charset.count {
                    result.append(charset[Int(random)])
                    remaining -= 1
                }
            }
        }
        return result
    }

    private static func sha256(_ input: String) -> String {
        let data = Data(input.utf8)
        let hash = SHA256.hash(data: data)
        return hash.map { String(format: "%02x", $0) }.joined()
    }
    #endif

    private func finishBootstrap(user: User, method: String) async throws {
        phase = .authenticatedBootstrapping
        let me = try await api.bootstrapMe()
        identity = ShellIdentity(
            userId: me.userId,
            displayName: me.displayName ?? user.displayName,
            email: me.email ?? user.email,
            firebaseUid: user.uid
        )
        MomentraIdentityCache.save(
            firebaseUid: user.uid,
            userId: me.userId,
            displayName: me.displayName ?? user.displayName,
            email: me.email ?? user.email
        )
        MomentraAnalytics.shared.syncUserDemographics(
            user: Auth.auth().currentUser,
            profileDisplayName: me.displayName ?? user.displayName,
            profileEmail: me.email ?? user.email
        )
        MomentraAnalytics.shared.trackAuthResult(method: method, success: true)
        phase = .authenticated
        error = nil
        APIClient.shared.warmAuthToken()
        PushNotifications.requestPermissionAndRegister()
        Task { await PushNotifications.syncDeviceWithBackend() }
    }

    private func recoverFromFailure(_ error: Error, method: String) {
        phase = .authError
        self.error = Self.userFacingAuthError(error)
        MomentraAnalytics.shared.trackAuthResult(
            method: method,
            success: false,
            errorCode: error.localizedDescription
        )
    }

    private static func userFacingAuthError(_ error: Error) -> String {
        if let kind = error as? APIErrorKind {
            switch kind {
            case .network(let message):
                return message
            case .unauthenticated:
                return "Session expired. Sign in again."
            case .server:
                return "The API is temporarily unavailable. Try again."
            default:
                return String(describing: kind)
            }
        }
        let ns = error as NSError
        if ns.domain == NSURLErrorDomain {
            return """
            Cannot reach the API at \(APIConfig.baseURLDescription). \
            Set Info.plist MomentraAPIBaseURL to your computer’s LAN IP \
            (same Wi‑Fi), matching Android API_BASE_URL. Backend must be running.
            """
        }
        // Firebase / Apple often localize this as “Could not connect to the server.”
        let text = error.localizedDescription
        if text.localizedCaseInsensitiveContains("could not connect")
            || text.localizedCaseInsensitiveContains("can't connect")
            || text.localizedCaseInsensitiveContains("network error") {
            return """
            Network error during sign-in. If Firebase succeeded but bootstrap failed, \
            check MomentraAPIBaseURL (\(APIConfig.baseURLDescription)) and that the \
            backend is listening on 0.0.0.0:3000.
            """
        }
        return text
    }

    func sendPasswordReset(email: String) {
        let trimmed = email.trimmingCharacters(in: .whitespacesAndNewlines)
        guard !trimmed.isEmpty else {
            error = "Enter your email to reset password."
            return
        }
        phase = .authenticating
        error = nil
        Auth.auth().sendPasswordReset(withEmail: trimmed) { err in
            Task { @MainActor in
                if let err {
                    self.phase = .authError
                    self.error = err.localizedDescription
                } else {
                    self.phase = .signedOut
                    self.error = "Password reset email sent. Check your inbox."
                }
            }
        }
    }

    func signOut() {
        MomentraIdentityCache.clear(firebaseUid: Auth.auth().currentUser?.uid)
        BootstrapCacheStore.clear(userId: identity?.userId)
        UserDefaults.standard.removeObject(forKey: "momentra_hide_balances")
        try? Auth.auth().signOut()
        #if os(iOS)
        GIDSignIn.sharedInstance.signOut()
        #endif
        identity = nil
        phase = .signedOut
    }

    func onSessionExpired() {
        MomentraIdentityCache.clear(firebaseUid: Auth.auth().currentUser?.uid)
        BootstrapCacheStore.clear(userId: identity?.userId)
        try? Auth.auth().signOut()
        #if os(iOS)
        GIDSignIn.sharedInstance.signOut()
        #endif
        identity = nil
        phase = .sessionExpired
        error = "Session expired"
    }
}

#if os(iOS)
extension UIApplication {
    static func topViewController(
        base: UIViewController? = UIApplication.shared.connectedScenes
            .compactMap { ($0 as? UIWindowScene)?.keyWindow }
            .first?.rootViewController
    ) -> UIViewController? {
        if let nav = base as? UINavigationController {
            return topViewController(base: nav.visibleViewController)
        }
        if let tab = base as? UITabBarController, let selected = tab.selectedViewController {
            return topViewController(base: selected)
        }
        if let presented = base?.presentedViewController {
            return topViewController(base: presented)
        }
        return base
    }
}
#endif
